File: Common/Json/Dynamic/JsonDynamicExtensions.vb
Imports System.Runtime.CompilerServices

Namespace Common

    Public Module JsonDynamicExtensions

        <Extension()>
        Public Sub WriteValue(ByVal Writer As JsonWriter, ByVal Obj As JsonDynamicBase, Optional ByVal MultiLine As Boolean? = Nothing)
            If TypeOf Obj Is JsonDynamicDictionary Then
                Dim T = DirectCast(Obj, JsonDynamicDictionary)

                Dim ML As Boolean
                If MultiLine.HasValue Then
                    ML = MultiLine.Value
                Else
                    ML = T.Count > 1
                    ML = ML OrElse T.Values.Any(Function(I) TypeOf I IsNot JsonDynamicValue)
                End If

                Using Writer.OpenDictionary(ML)
                    For Each I In T.OrderBy(Function(KV) KV.Key)
                        Writer.WriteKey(I.Key)
                        Writer.WriteValue(I.Value, MultiLine)
                    Next
                End Using

                Exit Sub
            End If

            If TypeOf Obj Is JsonDynamicList Then
                Dim T = DirectCast(Obj, JsonDynamicList)

                Dim ML As Boolean
                If MultiLine.HasValue Then
                    ML = MultiLine.Value
                Else
                    ML = T.Count > 10
                    ML = ML OrElse T.Any(Function(I) TypeOf I IsNot JsonDynamicValue)
                    ML = ML OrElse T.Cast(Of JsonDynamicValue).Sum(Function(I) I.Value.Length) > 150
                End If

                Using Writer.OpenList(ML)
                    For Each I In T
                        Writer.WriteValue(I, MultiLine)
                    Next
                End Using

                Exit Sub
            End If

            If TypeOf Obj Is JsonDynamicValue Then
                Dim T = DirectCast(Obj, JsonDynamicValue)
                Writer.WriteValue(T.Value, T.IsString)

                Exit Sub
            End If

            Verify.Fail("Invalid object type.")
        End Sub

        <Extension()>
        Public Function ToDynamic(ByVal Self As JsonObject) As JsonDynamicBase
            If TypeOf Self Is JsonDictionaryObject Then
                Dim Res = New JsonDynamicDictionary()
                For Each T In DirectCast(Self, JsonDictionaryObject)
                    Res.Add(T.Key, T.Value.ToDynamic())
                Next
                Return Res
            End If

            If TypeOf Self Is JsonListObject Then
                Dim Res = New JsonDynamicList()
                For Each T In DirectCast(Self, JsonListObject)
                    Res.Add(T.ToDynamic())
                Next
                Return Res
            End If

            If TypeOf Self Is JsonValueObject Then
                Dim T = DirectCast(Self, JsonValueObject)
                Return New JsonDynamicValue(T.Value, T.IsString)
            End If

            Verify.Fail("Invalid object type.")
            Return Nothing
        End Function

        <Extension()>
        Public Function ToConstant(ByVal Self As JsonDynamicBase) As JsonObject
            If TypeOf Self Is JsonDynamicDictionary Then
                Dim Res = New List(Of KeyValuePair(Of String, JsonObject))()
                For Each T In DirectCast(Self, JsonDynamicDictionary)
                    Res.Add(New KeyValuePair(Of String, JsonObject)(T.Key, T.Value.ToConstant()))
                Next
                Return New JsonDictionaryObject(Res)
            End If

            If TypeOf Self Is JsonDynamicList Then
                Dim Res = New List(Of JsonObject)()
                For Each T In DirectCast(Self, JsonDynamicList)
                    Res.Add(T.ToConstant())
                Next
                Return New JsonListObject(Res)
            End If

            If TypeOf Self Is JsonDynamicValue Then
                Dim T = DirectCast(Self, JsonDynamicValue)
                Return New JsonValueObject(T.Value, T.IsString)
            End If

            Verify.Fail("Invalid object type.")
            Return Nothing
        End Function

        <Extension()>
        Public Function AsList(ByVal Self As JsonDynamicBase) As JsonDynamicList
            Verify.NonNull(Self)
            Return DirectCast(Self, JsonDynamicList)
        End Function

        <Extension()>
        Public Function AsValue(ByVal Self As JsonDynamicBase) As JsonDynamicValue
            Verify.NonNull(Self)
            Return DirectCast(Self, JsonDynamicValue)
        End Function

        <Extension()>
        Public Function AsDictionary(ByVal Self As JsonDynamicBase) As JsonDynamicDictionary
            Verify.NonNull(Self)
            Return DirectCast(Self, JsonDynamicDictionary)
        End Function

        <Extension()>
        Public Function GetString(ByVal Self As JsonDynamicValue) As String
            Return Self.Value
        End Function

        <Extension()>
        Public Function GetBoolean(ByVal Self As JsonDynamicValue) As Boolean
            Verify.True(Self.Value = JsonDynamicValue.True Or Self.Value = JsonDynamicValue.False, "Value must be a boolean.")
            Return Self.Value = JsonDynamicValue.True
        End Function

        <Extension()>
        Public Function GetInteger(ByVal Self As JsonDynamicValue) As Integer
            Dim T = 0
            Verify.True(ParseInv.TryInteger(Self.Value, T), "Value must be an integer.")
            Return T
        End Function

        <Extension()>
        Public Function GetDouble(ByVal Self As JsonDynamicValue) As Double
            Dim T = 0.0
            Verify.True(ParseInv.TryDouble(Self.Value, T), "Value must be a number.")
            Return T
        End Function

    End Module

End Namespace

File: Common/Json/Dynamic/JsonDynamicExtensions.cs
using System.Collections.Generic;
using System.Linq;

namespace Ks.Common
{
    public static class JsonDynamicExtensions
    {
        public static void WriteValue(this JsonWriter Writer, JsonDynamicBase Obj, bool? MultiLine = default)
        {
            switch (Obj)
            {
                case JsonDynamicDictionary dic:
                {
                    bool ML;
                    if (MultiLine.HasValue)
                    {
                        ML = MultiLine.Value;
                    }
                    else
                    {
                        ML = dic.Count > 1;
                        ML = ML || dic.Values.Any(I => !(I is JsonDynamicValue));
                    }

                    using (Writer.OpenDictionary(ML))
                    {
                        foreach (var I in dic.OrderBy(KV => KV.Key))
                        {
                            Writer.WriteKey(I.Key);
                            Writer.WriteValue(I.Value, MultiLine);
                        }
                    }

                    return;
                }
                case JsonDynamicList list:
                {
                    bool ML;
                    if (MultiLine.HasValue)
                    {
                        ML = MultiLine.Value;
                    }
                    else
                    {
                        ML = list.Count > 10;
                        ML = ML || list.Any(I => !(I is JsonDynamicValue));
                        ML = ML || list.Cast<JsonDynamicValue>().Sum(I => I.Value.Length) > 150;
                    }

                    using (Writer.OpenList(ML))
                    {
                        foreach (var I in list)
                        {
                            Writer.WriteValue(I, MultiLine);
                        }
                    }

                    return;
                }
                case JsonDynamicValue v:
                {
                    Writer.WriteValue(v.Value, v.IsString);
                    return;
                }
            }

            Verify.Fail("Invalid object type.");
        }

        public static JsonDynamicBase ToDynamic(this JsonObject Self)
        {
            switch (Self)
            {
                case JsonDictionaryObject dic:
                {
                    var Res = new JsonDynamicDictionary();
                    foreach (var T in dic)
                    {
                        Res.Add(T.Key, T.Value.ToDynamic());
                    }

                    return Res;
                }
                case JsonListObject list:
                {
                    var Res = new JsonDynamicList();
                    foreach (var T in list)
                    {
                        Res.Add(T.ToDynamic());
                    }

                    return Res;
                }
                case JsonValueObject v:
                {
                    return new JsonDynamicValue(v.Value, v.IsString);
                }
            }

            Verify.Fail("Invalid object type.");
            return null;
        }

        public static JsonObject ToConstant(this JsonDynamicBase Self)
        {
            switch (Self)
            {
                case JsonDynamicDictionary dic:
                {
                    var Res = new List<KeyValuePair<string, JsonObject>>();
                    foreach (var T in dic)
                    {
                        Res.Add(new KeyValuePair<string, JsonObject>(T.Key, T.Value.ToConstant()));
                    }

                    return new JsonDictionaryObject(Res);
                }
                case JsonDynamicList list:
                {
                    var Res = new List<JsonObject>();
                    foreach (var T in list)
                    {
                        Res.Add(T.ToConstant());
                    }

                    return new JsonListObject(Res);
                }
                case JsonDynamicValue v:
                {
                    return new JsonValueObject(v.Value, v.IsString);
                }
            }

            Verify.Fail("Invalid object type.");
            return null;
        }

        public static JsonDynamicList AsList(this JsonDynamicBase Self)
        {
            Verify.NonNull(Self);
            return (JsonDynamicList) Self;
        }

        public static JsonDynamicValue AsValue(this JsonDynamicBase Self)
        {
            Verify.NonNull(Self);
            return (JsonDynamicValue) Self;
        }

        public static JsonDynamicDictionary AsDictionary(this JsonDynamicBase Self)
        {
            Verify.NonNull(Self);
            return (JsonDynamicDictionary) Self;
        }

        public static string GetString(this JsonDynamicValue Self)
        {
            return Self.Value;
        }

        public static bool GetBoolean(this JsonDynamicValue Self)
        {
            Verify.True((Self.Value == JsonDynamicValue.True) | (Self.Value == JsonDynamicValue.False), "Value must be a boolean.");
            return Self.Value == JsonDynamicValue.True;
        }

        public static int GetInteger(this JsonDynamicValue Self)
        {
            Verify.True(ParseInv.TryInteger(Self.Value, out var T), "Value must be an integer.");
            return T;
        }

        public static double GetDouble(this JsonDynamicValue Self)
        {
            Verify.True(ParseInv.TryDouble(Self.Value, out var T), "Value must be a number.");
            return T;
        }
    }
}
